db/migrate/20190906165501_remove_broken_variants_from_carts.rb

```diff
@@ -0,0 +1,17 @@
+class RemoveBrokenVariantsFromCarts < ActiveRecord::Migration
+  def up
+    # Removes line_items from open carts where the variant has a hard-deleted price
+
+    variants_without_prices = execute(
+      'SELECT spree_variants.id FROM spree_variants
+      LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id)
+      WHERE NOT EXISTS (SELECT 1 FROM spree_prices WHERE spree_prices.variant_id = spree_variants.id)'
+    ).to_a.map{ |v| v['id'].to_i }
+
+    broken_line_items = Spree::LineItem.
+      joins(:variant).where('spree_variants.id IN (?)', variants_without_prices).
```

Comment:
I know it might not make much difference given the numbers of affected variants but this is a pattern I see over and over again that has performance and complexity implications. We do a lot of pointless back and forth to the DB in cases that the DB can handself itself without having to resort to Ruby.

There's no need to collect the variant ids in line 9 to pass them back again if we perform a subquery. That can be achieved passing an ActiveRecord::Relation to the where method like:

```ruby
variants_without_prices = 'SELECT spree_variants.id FROM spree_variants
LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id)
WHERE NOT EXISTS (SELECT 1 FROM spree_prices WHERE spree_prices.variant_id = spree_variants.id)'

broken_line_items = Spree::LineItem.
      joins(:variant).
      joins(:order).
      where(variant: { id: variants_without_prices }).
      merge(Spree::Order.where(state: 'cart')).

broken_line_items.each(&:destroy)
```

Reply:
Updated :+1: